docs/enhancements/installation-progress-bar.md

```diff
@@ -0,0 +1,116 @@
+# Cluster installation progress bar
+
+## Goals
+
+* Progress should be computed in the BE by design and not in the UI
+  * that will also make it usable from other APIs as kube-api
+* Failing workers or OLM-operators shouldn't fail the cluster installation
+  * installed clusters with failing hosts or failing OLM-operators should still reach 100% progress.
+* We should align our design to other installation processes like OCM, ACM, etc.
+  * https://marvelapp.com/prototype/gfa4i5b/screen/80132841
+
+
+## Current implementation
+
+* Computed in UI
+* `progressPercent = hostsProgressPercent * 0.75 + operatorsProgressPercent * 0.25`
+  * `operatorsProgressPercent =(completedOperators / monitoredOperators) * 100`
+    * A completedOperator is one of the following:
+      * A built-in operator with `available` status
+      * An OLM operator with `available` or `failed` status
+  * `hostsProgressPercent = (ΣcompletedStages / ΣallStages) * 100`
+    * For each host the data is retrieved by using the following APIs
+
+```
+GET api/assisted-install/v1/clusters/{cluster_id} | jq '.hosts[0].progress'
+{
+  "current_stage": "Done",
+  "stage_started_at": "2021-05-31T13:37:26.224Z",
+  "stage_updated_at": "2021-05-31T13:37:26.224Z"
+}
+
+GET /api/assisted-install/v1/clusters/{cluster_id} | jq '.hosts[0].progress_stages'
+[
+  "Starting installation",
+  "Installing",
+  "Writing image to disk",
+  "Waiting for control plane",
+  "Rebooting",
+  "Configuring",
+  "Joined",
+  "Done"
+]
+```
+
+## New implementation
+
+* Will be computed in BE
+* `progres =ΣclusterStageProgress * Wi = preparingForInstallation * Wpi + installing * Wi + finalizing * Wf` while `ΣWi=1`
```

Comment:
Why not keep the current implementation logic and just move it to the backend?
Who will do all the progress calculations and when?
Will every host installation progress update trigger a calculation of the cluster progress? or it will get calculated and updated by some monitor?

Reply:
> Why not keep the current implementation logic and just move it to the backend?

Because the new implementation is more precise and more flexible.
By setting Wpi = 0, Wi = 0.75, Wf = 0.25 you get the current implementation anyway if that we want.

> Who will do all the progress calculations and when?

The BE, When is yet to be decided. Can be on API call, can be on refresh status etc.

> Will every host installation progress update trigger a calculation of the cluster progress? or it will get calculated and updated by some monitor?

Didn't decide yet, we can discuss it here. I think we can put it in the monitoring.

Reply:
So far the progress info didn't exist and we didn't need to keep it up to date.
I think that updating these new fields (host and cluster) will diminish the performance enhancements @tsorya and @ori-amizur recently added.

Reply:
Good point, but I think it should be OK:
1. During the actual installation the cluster resource is probably being modified anyway and we're not optimizing much if at all.
2. It would only need to be updated when a host finishes a stage, not constantly.

@tsorya and/or @ori-amizur should confirm.

Reply:
It may indeed.

@tsorya, @ori-amizur How bad do you guys think it will be to add those calculations to the monitoring processes?

What about computing it only on API call?
* On cluster API call:
  * trigger the hosts' progress computation
  * trigger the operators' progress computation
  * compute the total progress

* On Host API call trigger the hosts' computations only
* On Operator API call trigger the operators' computations only

Reply:
as i understand it will happened only during installation so shouldn't be a big deal

Reply:
I think adding the calculation to the monitor is not a good idea with regards to the performance improvements.  The logic of selecting the monitored entities might change soon, and we don't want progress change to trigger monitor on hosts or clusters. Computing it in API call seems more focused and consumes less resources.

Reply:
After talking to @ori-amizur this is how it will be handled:
* `preparing-for-installation` --> `installing` progress change will be handled by the monitoring process
  * The monitoring is the one responsible for changing that cluster stage.
* Hosts' stages progress will be updated by `PUT /clusters/{cluster_id}/hosts/{host_id}/progress` API
  * When this API is called to update the stage we will compute the new progress percentage as well.
* Operators' progress will be updated by the monitoring process
  * This is the only process that keeps track of the operators' statuses.

Reply:
> On Host API call trigger the hosts' computations only
Shouldn't the cluster progress be calculated as well?

Reply:
@filanov This is the updated solution decided https://github.com/openshift/assisted-service/pull/2042#discussion_r659296212.

Regarding the cluster progress I think the best is that each time an "internal" progress change, we compute the cluster total progress and store it as well.

This way when any of the API is called, we just need to fetch the data from the DB and return it to the user.